### PR TITLE
[red-knot] Cleanup generated names of mdtest tests

### LINKED ISSUE #13768: "Cleanup" markdown test names
I don't know if rstest allows to do that but the generated test names are hard to read and the enumerating of the tests also leads to instable identifiers across runs (changing test names can be very annoying when bisecting a test failures). 

image

It would be nice if we could:

* remove the `path` prefix
* remove the index
* remove the `resources_md` prefix
* remove the `md` postfix

Ideally, we would use the markdown title but that's probably not possible.

### PR DESCRIPTION
**Pull request by Aditya-PS-05**
Part of #13768 

## Summary
Added `dir-test` to `red_knot_python_semantic` and modify `mdtest` definition.

### CODE REVIEW

File: crates/red_knot_python_semantic/Cargo.toml
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/Cargo.toml:38]: Our preferred style for pinning dependencies is to put the pins in the `Cargo.toml` file in the repository root. So you'd put `dir-test = "0.3.0"` in the `Cargo.toml` file in the repository root, and in this `Cargo.toml` file you'd just put `dir-test = { workspace = true }`

File: crates/red_knot_python_semantic/tests/mdtest.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/tests/mdtest.rs:4]: Please add this doc-comment back -- it's very useful!

File: crates/red_knot_python_semantic/tests/mdtest.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/tests/mdtest.rs:22]: This certainly makes the tests run a lot faster! Unfortunately, that's because you're not actually testing anything anymore 😄 You need to make sure you're actually still calling `red_knot_test::run()` on the path after you've "cleaned the path up" :-)

**Review by AlexWaygood — CHANGES_REQUESTED** [review]
Thanks!

File: crates/red_knot_python_semantic/tests/mdtest.rs
Comment (Aditya-PS-05) [reply crates/red_knot_python_semantic/tests/mdtest.rs:22]: Sorry, I didn't understand. Are you trying to point out it's wrong and so I need to add `red_knot_test::run()`?

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/tests/mdtest.rs:22]: > Sorry, I didn't understand. Are you trying to point out it's wrong and so I need to add `red_knot_test::run()`?

Yes.

The way these tests work is a macro generates (at build time) a Rust test for each file in a specific directory. Each generated test runs `red_knot_test::run()` on a specific Markdown file here. The `red_knot_test::run()` call is the part of the test that actually makes assertions that red-knot is running correctly:

https://github.com/astral-sh/ruff/blob/7ca357119496ecae3bcc72549e0133e8b69c3ec6/crates/red_knot_python_semantic/tests/mdtest.rs#L13 

Your PR currently removes this call, meaning that the macro is generating a lot of tests, but those tests aren't really testing anything :-)

Comment (Aditya-PS-05) [reply crates/red_knot_python_semantic/tests/mdtest.rs]: Required changes done.
Thanks for review and sorry for blunders.

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/tests/mdtest.rs]: No worries!

**Review by AlexWaygood — APPROVED** [review]
Thanks! This LGTM. I confirmed it implements the feature @MichaReiser asked for, and that the tests still fail correctly if there's an incorrect assertion in a `.md` file. The tests also _appear_ to be running slightly faster for me locally as well, which is nice!

I'll wait for another approval before merging, though, as I'm pretty new to `dir-test`.

File: Cargo.lock
Comment (MichaReiser) [inline Cargo.lock:410]: Hmm, that's unfortunate that dir-test still uses `syn` one but so be it. We can always make a PR upstream

File: crates/red_knot_python_semantic/Cargo.toml
Comment (MichaReiser) [inline crates/red_knot_python_semantic/Cargo.toml:38]: This should be a dev dependency

**Review by MichaReiser — APPROVED** [review]
This looks good. Thanks. Let's move the `dir-test` dependency to the dev dependency section and we can land this

File: Cargo.toml
Comment (MichaReiser) [inline Cargo.toml:68]:
```suggestion
dir-test = { version = "0.3.0" }
```

File: crates/red_knot_python_semantic/Cargo.toml
Comment (MichaReiser) [inline crates/red_knot_python_semantic/Cargo.toml:46]:
```suggestion
ruff_db = { workspace = true, features = ["os", "testing"] }
ruff_python_parser = { workspace = true }
red_knot_test = { workspace = true }
red_knot_vendored = { workspace = true }

anyhow = { workspace = true }
dir-test = {workspace = true}
```

**Review by MichaReiser — APPROVED** [review]
Thanks

**Comment by Aditya-PS-05** [comment]
Can I ask a question? 
I am a student and I want to get an internship as in my college internship is necessary. Can you guys please suggest me advice or give me tips to get an work through open-source and how to improve my open-source work? 

Thanks

**Comment by AlexWaygood** [comment]
Hey @Aditya-PS-05. Thanks again for the PR! It was really useful, and an important improvement.

GitHub issue trackers and PR threads aren't really the best place to ask for this kind of advice. There are forums such as Reddit and various programming Discord servers where volunteers are often very happy to give this kind of advice, but on GitHub we prefer to be more focussed on specific coding projects and problems :-)

That said, my advice would be:
- Continue submitting open-source PRs! I managed to get hired on the back of my open-source experience, so it's definitely possible.
- Focus on PRs where you're confident you understand how to fix the problem. A PR that doesn't really address an issue or that approaches an issue in completely the wrong way can end up meaning that a maintainer has to spend more time reviewing and fixing the PR than they would have spent if they just did it themselves.

  This PR wasn't that -- I think it did save us some time! It did also have some important mistakes in it at the start, though, which I think could possibly have been avoided if you'd taken some more time to think about the problem before submitting the PR.
- Your ability to make excellent PRs will improve if you continue to improve your programming ability in other ways as well. Read lots of code, watch YouTube programming tutorials, read tutorials/howtos/StackOverflow questions, continue to work on your own hobby projects. You don't have to do all of these, but they're all great ways to continue learning and developing your skills, and they'll all make you a better open-source contributor in the long run.

Hope that helps. Good luck with your coding journey! :-)

**Pull request merged**
